### PR TITLE
Phase5 - change off-set based pagination to cursor based pagination

### PR DESCRIPTION
**Pull request by kanghowoo**
At first, let’s create SQL first and compare the performance between offset-based-pagination and cursor-based-pagination.

5-1 Mock implementation
Let’s create controller and service/storage interface for GET list of posts with cursor-based-pagination
And for service/storage implementations, please create mock implementation first.
Check your application API for GET working well and create PR

5-2 Use MySQL
Let’s implement storage implementation using MySQL.
Check your application API for GET list of posts with cursor-based-pagination working well with MySQL and create PR.

### CODE REVIEW

File: src/main/java/com/mide/gangsaeng/board/BoardServiceImpl.java [hunk 1]
```diff
@@ -38,4 +45,31 @@ public List<Board> getPage(int offset, int size) {
...
+
+        List<Board> page = boardRepository.getNextPage(request.getCursor(), size);
+        Cursor afterCursor = getCursor(page);
+
+        return new CursorBasedResponse<>(page, afterCursor);
+    }
+
+    private Cursor getCursor(List<Board> page) {
+        if (page.isEmpty()) {
+            throw new IllegalArgumentException("prev page is not present");
```
Comment (TK-one) [inline src/main/java/com/mide/gangsaeng/board/BoardServiceImpl.java:67]: Why did u throw error here?

File: src/main/java/com/mide/gangsaeng/board/BoardService.java [hunk 2]
```diff
@@ -2,8 +2,12 @@
...
 import java.util.List;
 
+import com.mide.gangsaeng.common.CursorBasedRequest;
+import com.mide.gangsaeng.common.CursorBasedResponse;
+
 public interface BoardService {
     void write(BoardRequest request);
     BoardResponse read(long id);
     List<Board> getPage(int offset, int size);
+    CursorBasedResponse<Board> getPageByCursor(CursorBasedRequest request, int size);
```
Comment (TK-one) [inline src/main/java/com/mide/gangsaeng/board/BoardService.java:12]: no need to fix
It's okay for now, but in this code, service layer should know about the detail of controller's response `CursorBasedResponse`. This is not good design but it's okay for toy project.

File: src/main/resources/mybatis/mappers/board/boardMapper.xml [hunk 3]
```diff
@@ -17,4 +17,19 @@
...
+    SELECT *
+    FROM boards
+    WHERE id > #{id}
+    ORDER BY id LIMIT #{size};
+  </select>
+
+  <select id="getNextPage" parameterType="HashMap" resultType="com.mide.gangsaeng.board.Board">
+    SELECT *
+    FROM boards
+    WHERE #{id} > id
```
Comment (TK-one) [inline src/main/resources/mybatis/mappers/board/boardMapper.xml:31]: oh... so confusing...
Can you fix this like
```suggestion
    WHERE id < #{id}
```

**Comment by TK-one** [comment]
Where is this part
>  At first, let’s create SQL first and compare the performance between offset-based-pagination and cursor-based-pagination.

Test this and record as comment in this PR

File: src/main/java/com/mide/gangsaeng/board/BoardServiceImpl.java (hunk 1 above)
Comment (kanghowoo) [reply src/main/java/com/mide/gangsaeng/board/BoardServiceImpl.java:67]: 맨 처음 페이지에서 존재하지 않는 이전 페이지를 클라이언트에게 혹여나 받게 될 경우를 대비해 작성함.

**Comment by kanghowoo** [comment]
This is a result of comparing two performances.
Execution environment : 
 DB : Mysql 8.2.0
 Table row (total) : 200,000
           column : 6

Cursor-based pagination is **18 times** faster than Offset-based pagination.

image

File: src/main/java/com/mide/gangsaeng/board/BoardServiceImpl.java (hunk 1 above)
Comment (TK-one) [reply src/main/java/com/mide/gangsaeng/board/BoardServiceImpl.java:67]: Then if you show 10 posts per page, and you have 10 posts in database.
If client click "next page" button, your server throws exception.

If you think in the client's perspective, how can they prevent to show "next page" if there is no more posts?

Comment (kanghowoo) [reply src/main/java/com/mide/gangsaeng/board/BoardServiceImpl.java]: Send 11 posts GET request, then check the number of results. (posts per page size + 1)
If the number is 11, it means the 'next page' is present. If the number is less than 11, it means this page is the end of the content.
We can notify the client that the 'next page' is not present by sending `null` in the next-cursor parameter.

File: src/main/java/com/mide/gangsaeng/board/BoardServiceImpl.java [hunk 4]
```diff
@@ -38,4 +42,48 @@ public List<Board> getPage(int offset, int size) {
...
+
+            page.sort((p1, p2) -> Long.compare(p2.getId(), p1.getId()));
+
+            if (page.size() == size) {
+                page.remove(0);
+                prevCursor = page.get(0).getId();
+            }
+
+            if (page.size() < pageSize) {
+                page = boardRepository.getNextPage(Long.MAX_VALUE, pageSize); //root page
```
Comment (TK-one) [inline src/main/java/com/mide/gangsaeng/board/BoardServiceImpl.java:63]: 그래도 이경우에도 nextCursor를 체크해야하지 않을까

Comment (Kyudong3) [reply src/main/java/com/mide/gangsaeng/board/BoardServiceImpl.java:63]: you guys awesome..

**Review by TK-one — APPROVED** [review]

**Pull request merged**